utils/terrain/src/map.rs

```diff
@@ -0,0 +1,120 @@
+/// requires `Ord` instead of `Hash`.
+#[derive(Clone, Debug)]
+pub struct VecMap<K, V> {
+    values: Vec<(K, V)>
+}
+
+impl<K, V> VecMap<K, V> {
+    /// Construct a new [`VecMap`]
+    pub fn new() -> Self {
+        VecMap::default()
+    }
+
+    /// The number of elements in the [`VecMap`]
+    pub fn len(&self) -> usize {
+        self.values.len()
+    }
+
+    /// Remove all elements from the [`VecMap`]
+    pub fn clear(&self) -> usize {
+        self.values.clear()
+    }
+
+    /// Reserve capacity for `additional` more elements to be inserted into
+    /// the [`VecMap`] to avoid frequent reallocations.
+    ///
+    /// See [`Vec::reserve()`] for more information.
+    pub fn reserve(&self, additional) -> usize {
+        self.values.reserve(additional)
+    }
+}
+
+impl<K: Ord, V> VecMap<K, V> {
+    /// Get the value associated with `key`, if it exists.
+    pub fn get(&self, key: &K) -> Option<&V> {
+        match self.values.binary_search_by(|k| k.0.cmp(&key)) {
+            Ok(found) => {
+                Some(&self.values[found].1)
+            }
+            Err(_) => {
+                None
+            }
+        }
+    }
+
+    /// Get the value associated with `key`, if it exists, as a mutable reference.
+    pub fn get_mut(&mut self, key: &K) -> Option<&mut V> {
+        match self.values.binary_search_by(|k| k.0.cmp(&key)) {
+            Ok(found) => {
+                Some(&mut self.values[found].1)
+            }
+            Err(_) => {
+                None
+            }
+        }
+    }
+
+
+    /// Insert `value` with `key`, returning the existing value if it exists.
+    pub fn insert(&mut self, key: K, value: V) -> Option<V> {
+        match self.values.binary_search_by(|k| k.0.cmp(&key)) {
+            Ok(found) => {
+                Some(mem::replace(&mut self.values[found].1, value))
+            }
+            Err(ins) => {
+                self.values.insert(ins, (key, value));
+                None
+            }
+        }
+    }
+
+    /// Remove the value at `key`, returning it if it exists.
+    pub fn remove(&mut self, key: &K) -> Option<V> {
+        match self.values.binary_search_by(|k| k.0.cmp(key)) {
+            Ok(found) => {
+                Some(self.values.remove(found).1)
+            }
+            Err(_) => {
+                None
+            }
+        }
+    }
+}
+
+impl<K, V> Default for VecMap<K, V> {
+    fn default() -> Self {
+        VecMap {
+            values: vec![]
+        }
+    }
+}
+impl<K: Ord, V> Index<&'_ K> for VecMap<K, V> {
+    type Output = V;
+    fn index(&self, key: &K) -> &V {
+        self.get(key).expect("VecMap could not find key")
+    }
+}
+impl<K: Ord, V> IndexMut<&'_ K> for VecMap<K, V> {
+    fn index_mut(&mut self, key: &K) -> &mut V {
+        self.get_mut(key).expect("VecMap could not find key")
+    }
+}
+
+
+impl<K, V> VecMap<K, V> {
+    pub fn iter(&self) -> impl Iterator<Item=(&K, &V)> {
```

Comment:
Is it useful to have a separate `iter_keys` and `iter_values`?

Reply:
Yes